builtin/credential/ldap/backend.go

```diff
@@ -101,106 +102,191 @@ func (b *backend) Login(req *logical.Request, username string, password string)
 			Scope:  2, // subtree
 			Filter: fmt.Sprintf("(%s=%s)", cfg.UserAttr, ldap.EscapeFilter(username)),
 		})
 		if err != nil {
-			return nil, logical.ErrorResponse(fmt.Sprintf("LDAP search for binddn failed: %v", err)), nil
+			return bindDN, logical.ErrorResponse(fmt.Sprintf("LDAP search for binddn failed: %v", err))
 		}
-		if len(sresult.Entries) != 1 {
-			return nil, logical.ErrorResponse("LDAP search for binddn 0 or not uniq"), nil
+		if len(result.Entries) != 1 {
+			return bindDN, logical.ErrorResponse("LDAP search for binddn 0 or not unique")
 		}
-		binddn = sresult.Entries[0].DN
+		bindDN = result.Entries[0].DN
 	} else {
 		if cfg.UPNDomain != "" {
-			binddn = fmt.Sprintf("%s@%s", EscapeLDAPValue(username), cfg.UPNDomain)
+			bindDN = fmt.Sprintf("%s@%s", EscapeLDAPValue(username), cfg.UPNDomain)
 		} else {
-			binddn = fmt.Sprintf("%s=%s,%s", cfg.UserAttr, EscapeLDAPValue(username), cfg.UserDN)
+			bindDN = fmt.Sprintf("%s=%s,%s", cfg.UserAttr, EscapeLDAPValue(username), cfg.UserDN)
 		}
 	}
-	if err = c.Bind(binddn, password); err != nil {
-		return nil, logical.ErrorResponse(fmt.Sprintf("LDAP bind failed: %v", err)), nil
-	}
 
-	userdn := ""
+	return bindDN, nil
+}
+
+func getUserDN(cfg *ConfigEntry,c *ldap.Conn, binddn string) (string , *logical.Response) {
+	userDN := ""
 	if cfg.UPNDomain != "" {
 		// Find the distinguished name for the user if userPrincipalName used for login
-		sresult, err := c.Search(&ldap.SearchRequest{
+		result, err := c.Search(&ldap.SearchRequest{
 			BaseDN: cfg.UserDN,
 			Scope:  2, // subtree
 			Filter: fmt.Sprintf("(userPrincipalName=%s)", ldap.EscapeFilter(binddn)),
 		})
 		if err != nil {
-			return nil, logical.ErrorResponse(fmt.Sprintf("LDAP search failed: %v", err)), nil
+			return userDN, logical.ErrorResponse(fmt.Sprintf("LDAP search failed for detecting user: %v", err))
 		}
-		for _, e := range sresult.Entries {
-			userdn = e.DN
+		for _, e := range result.Entries {
+			userDN = e.DN
 		}
 	} else {
-		userdn = binddn
+		userDN = binddn
 	}
 
-	var allgroups []string
-	var policies []string
-	resp := &logical.Response{
-		Data: map[string]interface{}{},
+	return userDN, nil
+}
+
+func getLdapGroups(cfg *ConfigEntry, c *ldap.Conn, userdn string, username string) ([]string, *logical.Response) {
+	// retrieve the groups in a string/bool map as a structure to avoid duplicates inside
+	ldapMap := make(map[string]bool)
+	// Fetch the optional memberOf property values on the user object
+	// This is the most common method used in Active Directory setup to retrieve the groups
+	result, err := c.Search(&ldap.SearchRequest{
+		BaseDN: userdn,
+		Scope:  0,        // base scope to fetch only the userdn
+		Filter: "(cn=*)", // bogus filter, required to fetch the userdn
+		Attributes: []string{
+			"memberOf",
+		},
+	})
+	if err != nil {
+		return nil, logical.ErrorResponse(fmt.Sprintf("LDAP fetch of distinguishedName=%s failed: %v", userdn, err))
+	}
+	if len(result.Entries) != 1 {
```

Comment:
What happens if there is no `memberOf` attribute? I would assume that in that case either `err` is not `nil` or `result.Entries` is empty, but it looks like if either of those is true, the function will exit before running the other searches later on.